pkg/controller/project/project_control.go

```diff
@@ -0,0 +1,319 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            *namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(project, gardenv1beta1.SchemeGroupVersion.WithKind(projectKind))},
+			Annotations:     namespaceAnnotations,
+			Labels:          namespaceLabels,
+		},
+	}
+	if _, err := c.k8sGardenClient.CreateNamespace(namespaceObj, true); err != nil {
+		message := fmt.Sprintf("Error while creating/updating namespace %s for project: %+v", namespaceObj.Name, err)
+		conditionProjectNamespaceReady = helper.ModifyCondition(conditionProjectNamespaceReady, corev1.ConditionFalse, gardenv1beta1.ProjectNamespaceCreationFailed, message)
+		projectLogger.Error(message)
+		c.updateProjectStatus(project, *conditionProjectNamespaceReady, *conditionProjectNamespaceEmpty, *conditionProjectShootsWithErrors)
+		return err
+	}
+	conditionProjectNamespaceReady = helper.ModifyCondition(conditionProjectNamespaceReady, corev1.ConditionTrue, gardenv1beta1.ProjectNamespaceReconciled, "Namespace has been reconciled.")
+
+	// Update namespace field in project status
+	if project.Status.Namespace == nil || len(*project.Status.Namespace) == 0 {
+		project.Status.Namespace = &namespaceObj.Name
+		project, err = c.updateProjectStatus(project)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create RBAC rules to allow project owner to read, update, and delete the project.
+	owners := []rbacv1.Subject{project.Spec.Owner}
+	// This is to make sure that the actual creator does not accidentally lock himself out of the project.
+	if createdBy, ok := project.Annotations[common.GardenCreatedBy]; ok && project.Spec.Owner.Name != createdBy {
+		owners = append(owners, rbacv1.Subject{
+			APIGroup: rbacv1.GroupName,
+			Kind:     rbacv1.UserKind,
+			Name:     createdBy,
+		})
+	}
+
+	values := map[string]interface{}{
+		"project": map[string]interface{}{
+			"name":   project.Name,
+			"uid":    project.UID,
+			"owners": owners,
+		},
+	}
+	if err := common.ApplyChart(c.k8sGardenClient, chartRenderer, filepath.Join(common.ChartPath, "garden-project", "charts", "project-rbac"), "project-rbac", namespaceObj.Name, values, nil); err != nil {
+		message := fmt.Sprintf("Error while creating RBAC rules %s for namespace: %+v", namespaceObj.Name, err)
+		conditionProjectNamespaceReady = helper.ModifyCondition(conditionProjectNamespaceReady, corev1.ConditionFalse, gardenv1beta1.ProjectNamespaceReconcileFailed, message)
+		projectLogger.Error(message)
+		c.updateProjectStatus(project, *conditionProjectNamespaceReady, *conditionProjectNamespaceEmpty, *conditionProjectShootsWithErrors)
+		return err
+	}
+
+	// Check ProjectNamespaceEmpty condition.
+	if namespaceEmpty {
+		conditionProjectNamespaceEmpty = helper.ModifyCondition(conditionProjectNamespaceEmpty, corev1.ConditionTrue, "NoShootsOrBackupInfrastructuresExist", fmt.Sprintf("Number of BackupInfrastructures = %d, Number of Shoots = %d", numberOfBackupInfrastructure, numberOfShoots))
+	} else {
```

Comment:
Should we handle this case using project-deletion-validation admission webhook? I might be overthinking here. my concern here is, this way project deletion is one time action. Once somebody marks the projects to be deleted for non-empty project. He will surely see project in terminating state but with condition `NamespaceDeletionImpossible`.  And suppose now he realised that there are active shoots on project and he want to continue working with it. After some time if owner deletes the shoot, project will be automatically vanished from system. But owner might don't get the reason why it go vanished, because of possibility that he forgets project deletion within that time period.  Above scenario might be more clear and makes sense if you think from dashboard perspective.

Reply:
I will not handle such a case. We cannot adapt our implementation to give the user the freedom to "forget" what he has done before.
However, we could enforce the same behaviour like we when someone wants to delete a Shoot: Before the `DELETE` call is accepted, the Shoot (Project) must be annotated with a special confirmation annotation.
What do you think about that?

Reply:
I'm fine with it. Annotation will provide safety step. 
It was just a concern, don't have any strong opinion about it.